p2panda-store-next/src/sqlite.rs

```diff
@@ -0,0 +1,490 @@
+    }
+}
+
+impl SqliteStoreBuilder {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    #[cfg(any(test, feature = "test_utils"))]
+    pub fn random_memory_url(mut self) -> Self {
+        // Combining Rust tests with in-memory databases can lead to unsound behaviour, this
+        // "workaround" assigns every temporary database a different, random name and keeps them
+        // isolated from other tests.
+        //
+        // See related issue: https://github.com/launchbadge/sqlx/issues/2510
+        self.url = format!(
+            "sqlite://dbmem{}?mode=memory&cache=private",
+            rand::random::<u32>()
+        );
+        self
+    }
+
+    pub fn database_url(mut self, url: &str) -> Self {
+        self.url = url.to_string();
+        self
+    }
+
+    pub fn max_connections(mut self, max_connections: u32) -> Self {
+        self.max_connections = max_connections;
+        self
+    }
+
+    pub fn create_database(mut self, create_database: bool) -> Self {
+        self.create_database = create_database;
+        self
+    }
+
+    pub fn run_default_migrations(mut self, run_migrations: bool) -> Self {
+        self.run_migrations = run_migrations;
+        self
+    }
+
+    pub async fn build<'a>(self) -> Result<SqliteStore<'a>, SqliteError> {
+        if self.create_database {
+            create_database(&self.url).await?;
+        }
+
+        let pool: sqlx::SqlitePool = SqlitePoolOptions::new()
+            .max_connections(self.max_connections)
+            .connect(&self.url)
+            .await?;
+
+        if self.run_migrations {
+            run_pending_migrations(&pool).await?;
+        }
+
+        Ok(SqliteStore::new(pool))
+    }
+}
+
+pub type Transaction<'a> = sqlx::Transaction<'a, Sqlite>;
+
+/// SQLite database with connection pool and transaction provider.
+///
+/// This struct can be cloned and used in multiple places in the application. Every cloned instance
+/// will re-use the same connection pool and have access to the same transaction instance if one
+/// was started. To guard against sharing transactions unknowingly across unrelated database
+/// queries, a concept of a `TransactionPermit` was introduced which does not protect from misuse
+/// but helps to make "holding" a transaction explicit.
+///
+/// Please note that SQLite strictly serializes transactions with _writes_ and will block any
+/// parallel attempt to begin another one. Processes starting a transaction will acquire a
+/// `TransactionPermit` and keep it until the transaction was committed or rolled back. If the
+/// query only involves _reads_ it is recommended to not use transactions and use the `execute`
+/// method directly as acquiring transactions will potentially block other processes to do work.
+///
+/// ## Design decisions
+///
+/// This storage API design was chosen to make the dynamics of the underlying SQLite database
+/// explicit. Internally any process can access the transaction object to do writes and
```

Comment:
One more sentence here succinctly explaining why exposing the underlying SQLite transaction is required could be good.